File: Python DSA/Arrays/container_with_most_water.py
"""
CONTAINER WITH MOST WATER - Medium
Given array height, find two lines that form container with most water.

Example: height = [1,8,6,2,5,4,8,3,7] → Output: 49

Logical Thinking:
1. Two pointers at both ends (widest container)
2. Calculate area = min(height[left], height[right]) * width
3. Move pointer with smaller height (potential for taller line)
4. Keep track of maximum area found

Key: Two pointers, greedy move shorter height
Time: O(n) | Space: O(1)
"""

def max_area(height):
    left, right = 0, len(height) - 1
    max_water = 0
    while left < right:
        width = right - left
        max_water = max(max_water, min(height[left], height[right]) * width)
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1
    return max_water


# Test cases
if __name__ == "__main__":
    print(f"Testing Container With Most Water...")
    print("Add your test cases here")
    print("✅ Refer to problem description for examples")

File: DSA/Two Pointers/Java/ContainerWithMostWater.java
// LeetCode 11: Container With Most Water
// You are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).

public class ContainerWithMostWater {
    public int maxArea(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int maxArea = 0;
        
        while (left < right) {
            int currentArea = Math.min(height[left], height[right]) * (right - left);
            maxArea = Math.max(maxArea, currentArea);
            
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        
        return maxArea;
    }
    
    public static void main(String[] args) {
        ContainerWithMostWater solution = new ContainerWithMostWater();
        int[] height1 = {1, 8, 6, 2, 5, 4, 8, 3, 7};
        int[] height2 = {1, 1};
        
        System.out.println("Max area: " + solution.maxArea(height1));
        System.out.println("Max area: " + solution.maxArea(height2));
    }
}